ports/esp32/README.ota.md

```diff
@@ -0,0 +1,405 @@
+		h = hashlib.sha256()
+		with open(path, 'rb') as f:
+			buf = None
+			while True:
+				buf = f.read(self.CHUNK_SIZE)
+				if not buf:
+					break
+				h.update(buf)
+		filehash = (ubinascii.hexlify(h.digest()).decode())
+		self.log.log('{} hash is "{}", should be "{}"'.format(path, filehash, hash), 'ota')
+		return filehash == hash
+
+	def deleteOldVFS(self):
+		deepDeleteFolder("/{}".format(self.nextbootpartition))
+		os.mkdir("/{}".format(self.nextbootpartition))
+
+	def copyVFS(self):
+		self.deleteOldVFS()
+		deepCopyFolder("/{}".format(self.curbootpartition), "/{}".format(self.nextbootpartition))
+
+	def unpackTar(self):
+		t = utar.TarFile(self.TAR_PATH)
+		updatebasepath = "/{}/".format(self.nextbootpartition)
+		for i in t:
+			gc.collect()
+			if i.type == utar.DIRTYPE:
+				os.mkdir(updatebasepath+i.name.replace('/',''))
+			else:
+				with open(updatebasepath+i.name, 'wb') as ef:
+					pf = t.extractfile(i)
+					copyfileobj(pf, ef)
+
+	def finishUpdate(self):
+		esp.ota_set_boot_partition(bytearray(self.partitions[self.nextbootpartition][6]))
+	        self.log.log("Update finished, restarting", 'ota')
+		machine.reset()
+```
+
+## Example of how to call the updater class
+```python
+from foo import Logger, Updater
+import urequests
+
+log = Logger()
+ota =  Updater(log)
+# WARNING: you MUSST secure this connection in productive code, else anybody who can intercept this traffic has arbitrary code execution on your device. You could for example use ssl with certificate pinning
+metadata = urequests.get('http://10.42.0.1:8000/updatemeta').json()
+ota.update(metadata)
+```
+
+## Example server side
+This server requires bottle, you can find it here: https://bottlepy.org/
+
+```python
+#!/usr/bin/env python3
+from bottle import route, run, response, request, url, debug
+import os
+import json
+import datetime
+import hashlib
+
+partfile = 'ota.img'
+vfsfile = 'ota.tar'
+otahost = 'http://10.42.0.1:8000'
+chunk_size = 512
+
+do_update = False
+
+@route('/updatemeta', method=['POST', 'GET'])
+def trackerprobe():
+	if do_update:
+		res['update'] = {}
+		res['update']['partition'] = {'url': '{}/chunk/part'.format(otahost), 'size': int(os.stat(partfile).st_size), 'hash': hashlib.sha256(open(partfile, 'rb').read()).hexdigest()}
+		res['update']['vfs'] = {'url': '{}/chunk/vfs'.format(otahost), 'size': int(os.stat(vfsfile).st_size), 'hash': hashlib.sha256(open(vfsfile, 'rb').read()).hexdigest()}
+	print(res)
+	return json.dumps(res)
+
+
+@route('/chunk/<type>/<id:int>', name='chunk_id')
+@route('/chunk/<type>/', name='chunk')
+def chunk(type, id=0):
+	path = None
+	if type == 'part':
+		path = partfile
+	elif type == 'vfs':
+		path = vfsfile
+	if path:
+		f = open(path, 'rb')
+		f.seek(chunk_size*id)
+		data = f.read(chunk_size)
+		f.close()
+		response.set_header('Content-type', 'application/octet-stream')
+		print('type {}, chunk {}'.format(type, id))
+		return data
+	else:
+		return None
+
+
+debug()
+run(host='0.0.0.0', port=8000)
+```
```

Comment:
The Python code style looks bad, isn't it?

Reply:
This is very small example server, it's fine for that but obviosly nobody should use that in production.